### DISCUSSION TITLE
Next steps after Python 2/3 compatibility release

### DISCUSSION THREAD

**Discussion by ross-spencer** [exponential-decay/demystify#68] [Ideas]
Unless something Da Vinci code happens to me in the next few hours tonight, all being well, I anticipate a Python 2 and 3 release of this code today. 

![](https://media.giphy.com/media/VdESmCbCX3mxUPP9Zi/giphy.gif)

It has been, quite something. 

![image](https://user-images.githubusercontent.com/1880412/149681364-8be8ffbf-2727-4a34-a426-f653bd8006a1.png)

While still relatively happy with the separation of concerns in the code, e.g. much of the effort was in adding tests to the extensive analysis module. The output mechanism is really everything you'd expect of a lone-programmer in their early years, 7 years prior to today, without code-review, or a mentor.

While everything (knock on wood) should still be more stable than previously, there's a lot to be done still. 

![](https://media.giphy.com/media/SWzXMmicVNZl6QdVLw/giphy.gif)

A short list of next steps and notes are jotted down below. Look at them as a free for all for anyone who wants to help. 

* Cherry-pick ignored messages from Pylint and work to remove the majority of them. These messages are less-serious than the ones I wanted to pick up before releasing, but it is still good to look at them, be that holistically, or a module by module basis: [pylintrc](https://github.com/exponential-decay/demystify/blob/5580184fd9e5ab6d8634cfc5f33c65d20220c82e/.pylintrc).
* Translations as is always on the list: [Translation link](https://docs.google.com/spreadsheets/d/1dVsRsXgD9V2GarNHHpf6Tzhrfx99_MXt3LjSSDrNLOY/edit?usp=sharing).
* Packaging sqlitefid and this: https://github.com/exponential-decay/demystify/issues/48
* JSON output of the analysis results class: [here](https://github.com/exponential-decay/demystify/blob/5580184fd9e5ab6d8634cfc5f33c65d20220c82e/libs/AnalysisResultsClass.py). This is probably going to be horrific, but if it works, of if there is a way to make the JSON look clean, it might be a good view. 
* Expose queries as an API: [here](https://github.com/exponential-decay/demystify/issues/54). The queries should all be very standard, and might be useful in their own right. More soon on this. 
* Testing and wider-use. If the tool is useful or you have ideas for analysis, let me know, or submit a PR. Adding space used by duplicates seems like a cool stat: https://github.com/exponential-decay/demystify/issues/44.
* Python 3 only. I'm not planning on removing extraneous code in one go, but as time moves on from this point, the focus will be Python 3 so if the work can be improved or simplified with anything on offer in Python 3, that's cool.
* Remove LXML as a dependency: https://github.com/exponential-decay/demystify/issues/50 - it still bugs me to this day we need this for the HTML output but it shouldn't be necessary. How much effort is required to get rid of it? 

If you don't see an issue that you want to work on, but have an idea, open an issue, describe it, and submit a PR next to it, with tests. Or if you need a hand with any thing, signal that in the PR or issue and I can help. 

I'll keep refining this post as time goes on too, and will keep y'all updated on other progress. 

Danke, merci, and thank you! 

![](https://media.giphy.com/media/fWfowxJtHySJ0SGCgN/giphy.gif)

**Comment by joshuatj** [comment]
Congratulations on achieving this important milestone!
